### ISSUE TITLE
cleanupJob parallelism

### ISSUE THREAD

**Issue by lrochette** [electric-cloud/EC-Admin#40]
Is it possible to increase the ability of the jobsCleanup to perform across multiple resources in parallel?  Or have multiple steps that work on a different portion of the data?    
Thanks!

Request by Richard Long

**Comment by tanaynagjee** [comment]
Whether we can parallelize depends on the bottlenecks.  When it comes to deleting jobs, the DB is often the bottleneck, as the server's background deleter works its way through the deleting the objects from the DB.  Can you try running two different instances in parallel on your system?  For example, if you can delete 100 jobs in 10 minutes, how long does it take for 2 parallel runs each deleting a separate set of 100 jobs?

**Comment by lrochette** [comment]
I don't think you can safely run 2 instances of jobsCleanup as the find will return (more or less) the same list of jobs and then will fail as the job is already deleted

Maybe the trick is to run it with different "pattern matching" to get a list of different jobs.

**Comment by semweblongshot** [comment]
Laurent, attempting to run 2 instances of jobsCleanup did exactly as you indicated, the second instance essentially errors out on the first delete attempt.

What I was thinking of was a way to partition the data and use multiple resources to perform the delete.  As it's deleting from the DB and the filesystem, I don't think that would cause too big of a back up.

**Comment by mjwesterhof** [comment]
In general, I agree that running two instances sorted by (for example) unique lists of projects is the simplest solution, and perhaps it may even be effective.

The best solution is to simply run the jobsCleanup often enough (perhaps even multiple times per day) so that it does not have a lot of work to do and thus the need for parallelism goes away.  This is the best solution quite simply because this is the lowest-impact solution -- anything that increases the load placed on your server or agents or the workspace filesystems is a potential negative with regard to other jobs in the system -- and cleanup is like garbage collection in Java, users just hate when that disrupts their attempts to get useful work done.

With that said, an attempt to do a finer-grained parallelism, where the jobsCleanup job would actually do something like fetch the list of things to clean and then distribute subsets of that list to multiple agents (via multiple jobSteps) for cleaning is certainly possible.  But, it may not have the desired effect.  The problem is that doing so may actually introduce more bottlenecks than it solves.  As Tanay noted above, if the DB is the bottleneck, then multiple agents beating on the server to add jobs to the delete lists (for the background deleter to clean up) is not going to make things any faster, and probably will end up slowing things down.  On the other hand, if the effort is bottlenecked on agents deleting workspaces that are no longer needed, then it largely depends on how "smart" the deleter is about distributing that work -- the code would need to be refactored from its current simple iteration (deleting both job DB record and workspace for a given job, then moving to the next) to a structure where it does something like this:

Get list of jobs to be processed
Determine full set of workspaces to be considered
Identify best (or only) agent that can process each workspace (local workspaces being the issue)
Generate list of workspaces broken up by agent
Create parallel jobSteps for each agent:
    Get list of jobs and workspaces this agent is to process
    Delete all workspaces
    If sucessful, delete jobs

The ordering above should be sufficient to minimize the risk of orphaned workspaces, while still increasing the parallelism of the system.  (But, I still don't like it -- the algorithm above will certainly delete things more quickly, and in the worst case it shouldn't delete things significantly more slowly, but in many cases it has a much higher potential to slow down or impact other users, something I don't believe maintenance tasks should be permitted to do!)

In summary - run the delete more often.  If that fails, then a simple change to run multiple deletions sorted by project or date/time range will be easiest to implement.  If that fails, then run the delete more often.

**Comment by semweblongshot** [comment]
Thanks Mike, excellent points.  In our case, it's from a failed handoff and the ops team not running the {{deleteJobs}} and instead just clearing out the filesystem for old jobs.  So, we will be pushing to restore this as a regular process.  As it is, working through almost 200k old jobs before starting our migration to 6.0 is on track to complete before we begin.

**Comment by semweblongshot** [comment]
Laurent, re-reading through this, I think your point on pattern matching might be a way to accomplish our overdue process.

**Comment by lrochette** [comment]
Rich, the pattern is a "negative" one, it prevents jobs with a specific pattern to returned by the search.
It would be easy enough to duplicate jobsCleanup and simply change the filter from:

  push (@filterList, {"propertyName" => "jobName",
                      "operator" => "notLike",
                      "operand1" => $jobPattern});

to

  push (@filterList, {"propertyName" => "jobName",
                      "operator" => "like",
                      "operand1" => $jobPattern});

Also if you do not have to delete the workspace, the deleteObjects procedure will be a lot of faster. It was developed by Tanay for customers for a lot of jobs to delete

**Comment by lrochette** [comment]
No change planed in the near future.
